### ISSUE TITLE
Problem with Embedding layer definition in C++

### ISSUE THREAD

**Issue by ibalampanis** [mlpack/mlpack#2434]
I face the following problem. I want to build an LSTM model to make Sentiment Analysis. In C++, I use this wonderful library. Firstly, to understand the architecture I approached the problem via python, and Keras advised from some tutorials. This [tutorial][1] had this code:
```
    from keras.models import Sequential
    from keras.layers import Dense, Embedding, LSTM, GRU
    from keras.layers.embeddings import Embedding

    EMBEDDING_DIM = 100
   
    print('Build model...')

    model = Sequential()
    model.add(Embedding(vocab_size, EMBEDDING_DIM, input_length=max_length())
    model.add(GRU(units=32, dropout=0.2, recurrent_dropout=0.2))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
```
In my code, in C++, I have written this block:
```
    model.Add<IdentityLayer<> >();
    model.Add<Embedding<> >(vocabSize, embedSize);
    model.Add<LSTM<> >(inputSize, lstmCells, maxRho);
    model.Add<Dropout<> >(0.5);
    model.Add<ReLULayer<> >();
    model.Add<Linear<> >(lstmCells, outputSize);
```
The output console gives me this error:

    error: conv_to(): given object can't be interpreted as a vector
    terminate called after throwing an instance of 'std::logic_error'
    what():  conv_to(): given object can't be interpreted as a vector

I cannot understand where is the error after several hours of debugging ...
Thank you

  [1]: https://towardsdatascience.com/machine-learning-word-embedding-sentiment-classification-using-keras-b83c28087456

**Comment by rcurtin** [comment]
Hey @ibalampanis, sorry for the slow response.  Can you share the full code?  Is there a use of `conv_to()` anywhere in your code?  Or, perhaps, are you using the `Lookup` layer somewhere?  (That's the only layer I see that uses `conv_to<>` internally.)

**Comment by ibalampanis** [comment]
My full code is here: https://www.github.com/ibalampanis/distributed-training-of-recurrent-neural-networks-by-fgm-protocol/tree/master/cpp%2Fmodels%2Frnn.cc . Thanks.

**Comment by rcurtin** [comment]
Thanks!  I don't see anything immediately wrong with it.  It looks like it would be hard for me to reproduce the issue since that's the entire project; is there any chance that you can provide a small self-contained example that I could build with g++?

Also, if you can provide a backtrace from gdb or similar, I might be able to take a guess at what the issue is. :+1:

**Comment by ibalampanis** [comment]
Yes, tomorrowI will provide you my gdb log. Thank you.

**Comment by rcurtin** [comment]
Thanks!  Be sure to compile with `-g` so that we get line numbers too. :)

**Comment by ibalampanis** [comment]
Hello, this is my full bt log.
[bt.log](https://github.com/mlpack/mlpack/files/4785254/bt.log)
Thanks a lot.

**Comment by rcurtin** [comment]
I think I can see a very quick workaround, which is to use a batch size of 1.  Can you try that and see if it works?  I'm digging into a better fix than that though; hope to have something put together soon. :+1:

**Comment by zoq** [comment]
You are absolutely correct, the layer only supports batchSize = 1 right now, will open an issue.

**Comment by rcurtin** [comment]
Actually, I am not sure if my thought was correct.  @ibalampanis can you say a little about how you are encoding your input data when `datasetType == nlp` ? https://github.com/ibalampanis/distributed-training-of-recurrent-neural-networks-by-fgm-protocol/blob/master/cpp/models/rnn.cc#L77

My question is, what is the shape of the input?  I *think* that what you are doing is passing in a dictionary encoding of each word at each time step.  For the `Embedding` layer, this should actually not be one-hot encoded---it should just be an `int` representing the index of the dictionary encoding.  So, e.g., word 3 should be encoded just as `[1]` and not `[0 0 0 1 0 0 ...]`.  If you are using a one-hot encoded representation, try not doing that, and I think that it will work.

In whichever case solves your problem, I'll definitely add some better error checking so that other users won't be similarly stumped. :)

**Comment by ibalampanis** [comment]
> Actually, I am not sure if my thought was correct. @ibalampanis can you say a little about how you are encoding your input data when `datasetType == nlp` ? https://github.com/ibalampanis/distributed-training-of-recurrent-neural-networks-by-fgm-protocol/blob/master/cpp/models/rnn.cc#L77
> 
> My question is, what is the shape of the input? I _think_ that what you are doing is passing in a dictionary encoding of each word at each time step. For the `Embedding` layer, this should actually not be one-hot encoded---it should just be an `int` representing the index of the dictionary encoding. So, e.g., word 3 should be encoded just as `[1]` and not `[0 0 0 1 0 0 ...]`. If you are using a one-hot encoded representation, try not doing that, and I think that it will work.
> 
> In whichever case solves your problem, I'll definitely add some better error checking so that other users won't be similarly stumped. :)

I make the preprocess [here](https://github.com/ibalampanis/distributed-training-of-recurrent-neural-networks-by-fgm-protocol/blob/master/python/amazon-reviews-preproc.py) . 

Thanks all of you for your replies!

**Comment by rcurtin** [comment]
If you've already done `word2vec` on the data, it seems maybe there would be no need for an `Embedding` layer then?

**Comment by ibalampanis** [comment]
Yes, sure. I found this solution, but it is not via mlpack.

**Comment by rcurtin** [comment]
Perhaps the `Embedding` layer works differently in other toolkits.  Try removing it and let us know if that solves the issue?

**Comment by mrityunjay-tripathi** [comment]
I have tried to fix the issue in PR #2398. @rcurtin Let me know your thoughts.
